pkg/framework/context/context.go

```diff
@@ -110,19 +116,28 @@ func (p *Provider) InboundTransportEndpoint() string {
 	return p.inboundTransportEndpoint
 }
 
+func (p *Provider) tryToHandle(svc service.InboundHandler, msg service.DIDCommMsgMap, myDID, theirDID string) error {
+	if err := p.messenger.HandleInbound(svc, msg, myDID, theirDID); err != nil {
+		return fmt.Errorf("messenger HandleInbound: %w", err)
+	}
+
+	_, err := svc.HandleInbound(msg, myDID, theirDID)
+
+	return err
+}
+
 // InboundMessageHandler return an inbound message handler.
 func (p *Provider) InboundMessageHandler() transport.InboundMessageHandler {
 	return func(message []byte, myDID, theirDID string) error {
-		msg, err := service.NewDIDCommMsgMap(message)
+		msg, err := service.ParseDIDCommMsgMap(message)
 		if err != nil {
 			return err
 		}
 
 		// find the service which accepts the message type
 		for _, svc := range p.services {
 			if svc.Accept(msg.Type()) {
-				_, err = svc.HandleInbound(msg, myDID, theirDID)
-				return err
```

Comment:
why is messenger re-processing the msg in each iteration of the loop?

Reply:
messenger does not re-processing it, there is the return statement.

Reply:
Ah. Right.

Although logically, I think it is more clear (to me) that messenger is a pre-processor prior to calling services. If we start doing service chains (calling more than one service), I assume messenger should only be called once.

Reply:
I thought the messenger would be the new inbound dispatcher?

Reply:
@troyronda  `If we start doing service chains` - it looks like middleware. If we implement it, the messenger should be first in the chain. But I don't know how this implementation will look like. The messenger signature function for handle inbound might be the same as a service handle inbound function. It means that messenger can play a role as a service (and should be first in the chain).